src/google/protobuf/conformance_protobufcc.cc

```diff
@@ -0,0 +1,164 @@
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include <errno.h>
+#include <stdarg.h>
+#include <unistd.h>
+
+#include "google/protobuf/conformance.pb.h"
+
+using std::string;
+using conformance::ConformanceRequest;
+using conformance::ConformanceResponse;
+using conformance::TestAllTypes;
+
+int test_count = 0;
+bool verbose = false;
+
+void FatalError(const char *fmt, ...) {
+  va_list args;
+  va_start(args, fmt);
+  vfprintf(stderr, fmt, args);
+  va_end(args);
+  exit(1);
+}
+
+bool CheckedRead(int fd, void *buf, size_t len) {
+  size_t ofs = 0;
+  while (len > 0) {
+    ssize_t bytes_read = read(fd, (char*)buf + ofs, len);
+
+    if (bytes_read == 0) return false;
+
+    if (bytes_read < 0) {
+      FatalError("Error reading from test runner: %s\n", strerror(errno));
+    }
+
+    len -= bytes_read;
+    ofs += bytes_read;
+  }
+
+  return true;
+}
+
+void CheckedWrite(int fd, const void *buf, size_t len) {
+  if (write(fd, buf, len) != len) {
+    FatalError("Error writing to test runner: %s\n", strerror(errno));
+  }
+}
+
+void DoTest(const ConformanceRequest& request, ConformanceResponse* response) {
+  TestAllTypes test_message;
+
+  switch (request.payload_case()) {
+    case ConformanceRequest::kProtobufPayload:
+      if (!test_message.ParseFromString(request.protobuf_payload())) {
+        // Getting parse details would involve something like:
+        //   http://stackoverflow.com/questions/22121922/how-can-i-get-more-details-about-errors-generated-during-protobuf-parsing-c
+        response->set_parse_error("Parse error (no more details available).");
+        return;
+      }
+      break;
+
+    case ConformanceRequest::kJsonPayload:
+      response->set_runtime_error("JSON input is not yet supported.");
+      break;
+
+    case ConformanceRequest::PAYLOAD_NOT_SET:
+      FatalError("Request didn't have payload.\n");
+      break;
+  }
+
+  switch (request.requested_output()) {
+    case ConformanceRequest::UNSPECIFIED:
+      FatalError("Unspecified output format");
+      break;
+
+    case ConformanceRequest::PROTOBUF:
+      test_message.SerializeToString(response->mutable_protobuf_payload());
+      break;
+
+    case ConformanceRequest::JSON:
+      response->set_runtime_error("JSON output is not yet supported.");
+      break;
+  }
+}
+
+bool DoTestIO() {
```

Comment:
Google Code Style suggests DoTestIo instead of DoTestIO (the last o needs be lower-cased).